leker - Technical training

lkerroum opened the pull request.

lkerroum committed:
[ADD] estate: init

[IMP] manifest: author and license

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/1063.png)](https://mergebot.odoo.com/odoo/tutorials/pull/1063)

lkerroum committed:
[IMP] estate: real estate properties model definition

lkerroum committed:
[IMP] estate: added access rights

lkerroum committed:
[IMP] estate: added new record action and menu

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,21 @@
+    garden = fields.Boolean('Garden')
+    garden_area = fields.Integer('Garden area')
+    garden_orientation = fields.Selection(string='Garden Orientation',
+                                          selection=[('north', 'North'), ('south', 'South'), ('est', 'Est'), ('west', 'West')])
```
I prefer that way to be more readable for the selection fields
```suggestion
    garden_orientation = fields.Selection([
        ('north', 'North'),
        ('south', 'South'),
        ('east', 'East'),
        ('west', 'West')
    ], string='Garden Orientation',)
```

lkerroum replied:
Done! Thanks

amah-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,6 @@
+'application': True,
+'author': 'leker',
+'license': 'LGPL-3',
+}
```
let's go with this format 👀 
```suggestion
{
    'name': 'Real Estate',
    'depends': ['base'],
    'application': True,
    'author': 'leker',
    'license': 'LGPL-3',
}
```

lkerroum replied:
Done!

amah-odoo reviewed (commented):
Nice work 💪🏼 

Let's go with your first review 🎉

lkerroum committed:
[IMP] estate: added new fields and attributes to property model

amah-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,10 @@
+    'depends': ['base'],
+    'data': ['security/ir.model.access.csv',
+            'views/estate_property_views.xml',
+            'views/estate_menus.xml'],
```
let's be coherent with the rest of odoo modules 👀 
```suggestion
    'data': [
        'security/ir.model.access.csv',
        'views/estate_property_views.xml',
        'views/estate_menus.xml'
    ],
```

amah-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,10 @@
+            'views/estate_property_views.xml',
+            'views/estate_menus.xml'],
+    'application': True,
+    'author': 'leker',
```
The author is always `'Odoo S.A.'` for the RD because to develop for odoo it is not out own apps 🤔

amah-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,8 @@
+        <field name="res_model">estate.property</field>
+        <field name="view_mode">list,form</field>
+    </record>
+</odoo>
```
Do not forget to add the end of file empty line

lkerroum committed:
[IMP] estate: added list, form and search views

[FIX] estate: available filter bug

lkerroum committed:
[IMP] estate: new models property type, property tag, offer

amah-odoo commented on estate/views/estate_property_offer_views.xml:
```diff
@@ -0,0 +1,14 @@
+            </list>
+        </field>
+    </record>
+</odoo>
```
end of file new line ⚠️

amah-odoo commented on estate/models/__init__.py:
```diff
@@ -1 +1,6 @@
+    estate_property_offer,
+    estate_property_tag,
+    estate_property_type,
+)
```
I know that Ruff is happy with that but we actually prefer to have each import on its own line like 
```python
from . import estate_property
from . import estate_property_offer
...
```

amah-odoo commented on estate/views/estate_menus.xml:
```diff
@@ -0,0 +1,12 @@
+            <menuitem id="property_tags_model_menu_action" action="estate_property_tag_action"/>
+        </menuitem>
+    </menuitem>
+</odoo>
```
end of file new line ⚠️

amah-odoo reviewed (commented):
Thanks for your work so far 💪🏼

lkerroum committed:
[IMP] estate: added computed and onchange fields to property and offer models

lkerroum committed:
[IMP] estate: buttons for property and offer status

lkerroum committed:
[IMP] estate: model constraints

[FIX] estate: estate property style

[FIX] estate: found use of self inside a loop on self self.price

lkerroum committed:
[IMP] estate: enhance UI with widgets, inline views, and conditional buttons

[FIX] estate: domain ative_id bug

[FIX] estate: property type views prettier

[FIX] estate: missing color view in tag

[FIX] estate: always invisible field status removed from the offer view

lkerroum committed:
[IMP] estate: python and model inheritance

lkerroum committed:
[ADD] estate_account: module linked to estate, account, to create invoices

lkerroum committed:
[IMP] estate: added kanban view

lkerroum committed:
[FIX] estate: coding guiedlines

amah-odoo commented on estate/models/__init__.py:
```diff
@@ -0,0 +1,5 @@
+from . import estate_property_offer
+from . import estate_property_tag
+from . import estate_property_type
+from . import base_res_users
```
for the extension inherit using `_inherit` we advice to use same exact file name for file_name, ClassName  to ease the process of searching across all the extensions

amah-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,63 @@
+        return True
+
+    @api.model_create_multi
+    def create(self, vals):
```
as create can be done in batch is should take `vals_list` instead of `vals` it will be a list of dicts and each dict can have different values so you need to adapt the method 🤔

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,100 @@
+    def _check_selling_price_percentage(self):
+        for record in self:
+            if record.state == 'offer accepted' and float_compare(record.selling_price, 0.9 * record.expected_price, precision_digits=9) == -1:
+                msg = "The selling price cannot be lower than 90% of the expected price"
+                raise ValidationError(msg)
```
I believe error with raised here in the runbot when you have the string directly. this because the Runbot expect a translation object as error msg. we should discuss this together 🫡 
```suggestion
                raise ValidationError(self,env._("The selling price cannot be lower than 90% of the expected price"))
```

amah-odoo reviewed (commented):
Here is the last touches on this PR good luck with JS

Good job 👏🏼
